From 5790e730cf7d606478469b8671da7b5ee169044d Mon Sep 17 00:00:00 2001
From: Jiri Kosina <jkosina@suse.cz>
Date: Wed, 3 Jan 2018 15:20:04 +0100
Subject: [PATCH 30/64] kaiser: x86: Fix NMI handling

On Mon, 4 Dec 2017, Hugh Dickins wrote:

> kaiser-3.18.72.tar

Hi Hugh,

this hunk from 0024-kaiser-merged-update.patch:

	-       SWITCH_KERNEL_CR3_NO_STACK
	+       /*
	+        * percpu variables are mapped with user CR3, so no need
	+        * to switch CR3 here.
	+        */
	        cld
	        movq    %rsp, %rdx
	        movq    PER_CPU_VAR(kernel_stack), %rsp

is problematic, as the patchset actually never user-maps kernel_stack
percpu variable, and therefore crashes on NMIs.

The patch below is needed to make NMIs work properly.

Signed-off-by: Ben Hutchings <ben@decadent.org.uk>
Signed-off-by: Liu Haitao <haitao.liu@windriver.com>
---
 arch/x86/kernel/cpu/common.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/kernel/cpu/common.c b/arch/x86/kernel/cpu/common.c
index 07da020..b72ec06 100644
--- a/arch/x86/kernel/cpu/common.c
+++ b/arch/x86/kernel/cpu/common.c
@@ -1148,7 +1148,7 @@ DEFINE_PER_CPU(struct task_struct *, current_task) ____cacheline_aligned =
 	&init_task;
 EXPORT_PER_CPU_SYMBOL(current_task);
 
-DEFINE_PER_CPU(unsigned long, kernel_stack) =
+DEFINE_PER_CPU_USER_MAPPED(unsigned long, kernel_stack) =
 	(unsigned long)&init_thread_union - KERNEL_STACK_OFFSET + THREAD_SIZE;
 EXPORT_PER_CPU_SYMBOL(kernel_stack);
 
-- 
1.7.9.5

